Installation issue: mpich: undefined reference to `hwloc_topology_set_io_types_filter'

### Steps to reproduce the issue

I am trying to install firedrake on a cluster and would like to compile a new mpi. Part of the concretization results are listed below:
```console
 -   xffgfev          ^mpich@4.0.2%gcc@9.5.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci~rocm+romio~slurm~two_level_namespace~verbs+wrapperrpath datatype-engine=auto device=ch3 netmod=ofi pmi=pmi arch=linux-rhel6-ivybridge
[+]  yyr6dvc              ^findutils@4.9.0%gcc@9.5.0 patches=440b954 arch=linux-rhel6-ivybridge
[+]  i7hw5xi              ^hwloc@2.7.1%gcc@9.5.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm+shared arch=linux-rhel6-ivybridge
[+]  rjxaj6q                  ^libpciaccess@0.16%gcc@9.5.0 arch=linux-rhel6-ivybridge
[+]  otnpivf                      ^libtool@2.4.7%gcc@9.5.0 arch=linux-rhel6-ivybridge
[+]  f5lis6l                          ^m4@1.4.19%gcc@9.5.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-rhel6-ivybridge
[+]  edvnqxs                              ^libsigsegv@2.13%gcc@9.5.0 arch=linux-rhel6-ivybridge
[+]  edinkzp                      ^util-macros@1.19.3%gcc@9.5.0 arch=linux-rhel6-ivybridge
[+]  wenbrq5                  ^libxml2@2.9.13%gcc@9.5.0~python arch=linux-rhel6-ivybridge
[+]  bonjmeb                      ^xz@5.2.5%gcc@9.5.0~pic libs=shared,static arch=linux-rhel6-ivybridge
 -   tmynx25              ^libfabric@1.6.2%gcc@9.5.0~debug~kdreg fabrics=sockets,tcp,udp arch=linux-rhel6-ivybridge
```

In the building process, the linker will always reach to `/usr/lib64/libhwloc.so` other than the version installed by spack.
It seems the option `--with-hwloc-prefix` does not work. Even if I replace it with two other options: `-with-hwloc-include` and `--with-hwloc-lib`, the problem remains.

After I add the corresponding path of `hwloc` into `CFLAGS` and `LDFLAGS` in mpich's file `package.py`, the problem is gone.

@raffenet @yfguo, could you help find the reason? Thanks in advance!

### Error message

<details><summary>Error message</summary><pre>
/bin/sh ./libtool  --tag=CC   --mode=link /vol7/home/lrtfm/opt/spack/lib/spack/env/gcc/gcc  -g  -L.  -o mpiexec.hydra ui/mpich/mpiexec.o ui/mpich/utils.o -lpm -lhydra /tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/mpl/libmpl.la   -lpthread  -lhwloc
libtool: link: /vol7/home/lrtfm/opt/spack/lib/spack/env/gcc/gcc -g -o mpiexec.hydra ui/mpich/mpiexec.o ui/mpich/utils.o  -L. /tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/.libs/libpm.a /tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/.libs/libhydra.a /tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/mpl/.libs/libmpl.a -L/vol7/home/lrtfm/opt/spack/opt/spack/linux-rhel6-ivybridge/gcc-9.5.0/libfabric-1.14.1-2lx2gc4bktgpxgvkf3z2csotw6sq2wfc/lib -lpthread /usr/lib64/libhwloc.so -lm -lnuma -lxml2
/vol7/home/lrtfm/opt/spack/opt/spack/linux-rhel6-ivybridge//gcc-9.2.0vol7/binutils/-2.38home-/royhsidjkisjn7tmarni7kjty6n57k2i/lrtfmbin//ldopt:/ spack//tmpopt//lrtfm/spackspack-/stage/linuxspack--stagerhel6-mpich--4.0.2ivybridge-/konutm24znsihjf4vvbxugdr6rdwzumy/gccspack--src9.2.0/src//binutilspm/-hydra2.38/-.libsroyhsidjkisjn7tmarni7kjty6n57k2i//libhydra.a(bintopo_hwloc.o)/:ld :in  function/ `tmpHYDT_topo_hwloc_init/':lrtfm
//tmpspack/lrtfm-/stagespack/-stagespack/spack--stagestage--mpich-mpich4.0.2--konutm24znsihjf4vvbxugdr6rdwzumy/4.0.2spack--src/konutm24znsihjf4vvbxugdr6rdwzumysrc//pmspack/-hydra/srctools//toposrc/hwloc//topo_hwloc.cpm:/577:hydra undefined/ .reference libsto/ `libhydra.ahwloc_topology_set_io_types_filter('topo_hwloc.o
): in function `HYDT_topo_hwloc_init':
/tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/tools/topo/hwloc/topo_hwloc.c:577: undefined reference to `hwloc_topology_set_io_types_filter'
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[3]: *** [hydra_nameserver] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [hydra_persist] Error 1
/vol7/home/lrtfm/opt/spack/opt/spack/linux-rhel6-ivybridge/gcc-9.2.0/binutils-2.38-royhsidjkisjn7tmarni7kjty6n57k2i/bin/ld: /tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/.libs/libhydra.a(topo_hwloc.o): in function `HYDT_topo_hwloc_init':
/tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/tools/topo/hwloc/topo_hwloc.c:577: undefined reference to `hwloc_topology_set_io_types_filter'
collect2: error: ld returned 1 exit status
make[3]: *** [hydra_pmi_proxy] Error 1
/vol7/home/lrtfm/opt/spack/opt/spack/linux-rhel6-ivybridge/gcc-9.2.0/binutils-2.38-royhsidjkisjn7tmarni7kjty6n57k2i/bin/ld: /tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/.libs/libhydra.a(topo_hwloc.o): in function `HYDT_topo_hwloc_init':
/tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra/tools/topo/hwloc/topo_hwloc.c:577: undefined reference to `hwloc_topology_set_io_types_filter'
collect2: error: ld returned 1 exit status
make[3]: *** [mpiexec.hydra] Error 1
make[3]: Leaving directory `/tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src/src/pm/hydra'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/tmp/lrtfm/spack-stage/spack-stage-mpich-4.0.2-konutm24znsihjf4vvbxugdr6rdwzumy/spack-src'
make: *** [all] Error 2

</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (6a240c9b392d9d519b5df14943732bd1831d25bf)
* **Python:** 3.8.13
* **Platform:** linux-rhel6-ivybridge
* **Concretizer:** clingo


### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/9220278/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/9220280/spack-build-out.txt)


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/mpich/package.py
+++ b/var/spack/repos/builtin/packages/mpich/package.py
@@ -280,7 +280,7 @@ with '-Wl,-commons,use_dylibs' and without
         for exe in exes:
             variants = []
             output = Executable(exe)(output=str, error=str)
-            if re.search(r"--with-hwloc[-prefix]*=embedded", output):
+            if re.search(r"--with-hwloc(-prefix)*=embedded", output):
                 variants.append("~hwloc")
 
             if re.search(r"--with-pm=hydra", output):

--- a/var/spack/repos/builtin/packages/mpich/package.py
+++ b/var/spack/repos/builtin/packages/mpich/package.py
@@ -457,15 +457,24 @@ with '-Wl,-commons,use_dylibs' and without
         config_args = [
             "--disable-silent-rules",
             "--enable-shared",
-            "--with-hwloc-prefix={0}".format(
-                spec["hwloc"].prefix if "^hwloc" in spec else "embedded"
-            ),
             "--with-pm={0}".format("hydra" if "+hydra" in spec else "no"),
             "--{0}-romio".format("enable" if "+romio" in spec else "disable"),
             "--{0}-ibverbs".format("with" if "+verbs" in spec else "without"),
             "--enable-wrapper-rpath={0}".format("no" if "~wrapperrpath" in spec else "yes"),
             "--with-yaksa={0}".format(spec["yaksa"].prefix if "^yaksa" in spec else "embedded"),
         ]
+
+        # hwloc configure option changed in 4.0
+        if spec.satisfies("@4.0:"):
+            config_args.append(
+                "--with-hwloc={0}".format(spec["hwloc"].prefix if "^hwloc" in spec else "embedded")
+            )
+        else:
+            config_args.append(
+                "--with-hwloc-prefix={0}".format(
+                    spec["hwloc"].prefix if "^hwloc" in spec else "embedded"
+                )
+            )
 
         if "~fortran" in spec:
             config_args.append("--disable-fortran")

--- a/var/spack/repos/builtin/packages/mpich/package.py
+++ b/var/spack/repos/builtin/packages/mpich/package.py
@@ -280,7 +280,7 @@ with '-Wl,-commons,use_dylibs' and without
         for exe in exes:
             variants = []
             output = Executable(exe)(output=str, error=str)
-            if re.search(r"--with-hwloc-prefix=embedded", output):
+            if re.search(r"--with-hwloc[-prefix]*=embedded", output):
                 variants.append("~hwloc")
 
             if re.search(r"--with-pm=hydra", output):

